### PR TITLE
Switch the math renderer from KaTeX to MathJax

### PR DESCRIPTION
**Pull request by benchristel**
## Summary:
This is accomplished by adapting the MathJaxRenderer interface to look
like KaTeX, since that's what Khan Exercises expects.

Issue: https://khanacademy.atlassian.net/browse/LC-1002

Test plan:

- Start the dev server: `./server.sh`.
- Visit http://localhost:8765
- View all the example graphies by selecting them from the dropdown.
  They should all look pretty much the same as the renderings on
  production (http://graphie-to-png.khanacademy.systems/)
- Click the "Convert to Image" button. Copy the URL that appears.
- Paste the URL back into a new graphie editor window. You should get the same rendered image and labels.
- Paste the URL into a Markdown image link in the [exercise editor](https://www.khanacademy.org/devadmin/content/exercises/interpreting-expressions/x348258ef/x53040f6b0fa6ee1e), e.g. ``. The image and labels should render correctly in the preview.

Screen Shot 2023-11-30 at 4 28 02 PM

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -13,3 +13,6 @@
 [submodule "third_party/rappar"]
 	path = third_party/rappar
 	url = https://github.com/DmitryBaranovskiy/rappar.git
+[submodule "mathjax-renderer"]
+	path = mathjax-renderer
+	url = git@github.com:khan/mathjax-renderer
```
Comment (Myranae) [inline .gitmodules:18]: Question: The formatting here is different than the other urls; could that cause any problems? I know I have seen this format before, but just wanted to check.

**Review by Myranae — COMMENTED** [review]
When I tried looking at them side-by-side, I got an error. 
image
 Also added a non-blocking question in-line.

File: .gitmodules (hunk 1 above)
Comment (benchristel) [reply .gitmodules:18]: This format is needed because the mathjax-renderer repo is private. It's not possible to clone it using `https`, since you can't authenticate to GitHub over HTTPS. If you clone with SSH, though, GitHub can identify you by your SSH key, so it knows you have access.

**Comment by Myranae** [comment]
Got an error when trying to convert to an image
image
More details
image

**Comment by benchristel** [comment]
I can repro the INTERNAL SERVER ERROR on my dev machine. The issue seems to be that the `phantomjs` NPM package isn't executable or isn't installed correctly. Note that I had to run `npm install` to get the packages. But even after doing that, I get this error:

```
[2023-12-01 10:32:38 -0800] [56407] [DEBUG] POST /svg
internal/child_process.js:408
    throw errnoException(err, 'spawn');
    ^

Error: spawn Unknown system error -86
    at ChildProcess.spawn (internal/child_process.js:408:11)
    at spawn (child_process.js:553:9)
    at Object.<anonymous> (/Users/bchristel/khan/graphie-to-png/node_modules/phantomjs/bin/phantomjs:22:10)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47 {
  errno: 'Unknown system error -86',
  code: 'Unknown system error -86',
  syscall: 'spawn'
}
```

Digging into the phantomjs code, I don't see how it's supposed to work. It looks like maybe it's executing the wrong file? I'm not sure.

I don't think switching to MathJax is likely to be the cause of this error. One way to be sure is to use `https://github.com/Khan/internal-services/tree/master/graphie-to-png` to run the graphie editor in a Docker container—that's a better replica of what will run in production. But it also requires that you have Docker Desktop installed. I will try it on my machine.

**Comment by benchristel** [comment]
Re: the INTERNAL SERVER ERROR issue: I _was_ able to repro when running this in a Docker container. I could also repro it when running a new docker image built from the latest `master`. However, a container/image I created a month ago from `master` doesn't have the bug.

The issue is likely to be that our docker build process doesn't completely lock down the versions of all dependencies, so it's getting different code now than it was a month ago. **We can't deploy until that's fixed,** because it would break production.

However, because the issue repros on a new docker image built from `master`, I'm convinced that the error isn't due to anything I changed in this PR. @Myranae what do you think?

**Comment by Myranae** [comment]
Hmmm, it doesn't seem like the issue is with your code, but I think it might be better to wait to merge until we are able to get it working as intended. I'm not sure what the best option is, but if we can't get it working correctly, then I would worry about not working code being in `master`. Graphie-to-png is deployed using Docker, right? I would probably figure out if there is work going on to fix our Docker build issue, make this blocked by that one, and keep it ready for testing (if we are the ones who can actually fix it that is). That might not be the best practices approach though.

File: .gitmodules (hunk 1 above)
Comment (jeremywiebe) [reply .gitmodules:18]: Can we not use `package.json` to define this dependency?  Although, maybe it's all moot given we don't plan to maintain/upgrade this project much more.

File: templates/plain_graph.html [hunk 2]
```diff
@@ -65,14 +67,14 @@
...
 {% set KATHJAX_HASH = "eaf1d9cafdc2fc68897cf72d30d67d5a" %}
 
+<script src="static/js/mathjax-renderer/dist/mathjax-renderer.min.js"></script>
+<script src="static/js/mathjax-adapter.js"></script>
 <script src="static/js/khan-exercises/local-only/jquery.js"></script>
 <script src="static/js/khan-exercises/local-only/jed.js"></script>
 <script src="static/js/khan-exercises/local-only/i18n.js"></script>
 <script src="static/js/khan-exercises/local-only/localeplanet/icu.en.js"></script>
 <script src="static/js/khan-exercises/local-only/underscore.js"></script>
-<script src="static/js/khan-exercises/third_party/MathJax/2.1/MathJax.js?config=KAthJax-{{ KATHJAX_HASH }}"></script>
```
Comment (jeremywiebe) [inline templates/plain_graph.html:73]: 🥳  (also happy to see we're moving away from this really complex/confusing config mechanism!)

**Comment by benchristel** [comment]
tl;dr, I didn't quite figure out the PhantomJS issue, but I did realize that we don't actually need to use PhantomJS at all. So I deleted it. @Myranae generating the image should work now.

Details:

PhantomJS is basically a way to run a headless browser (i.e. a browser without a GUI window) on the server. I think the reason we were using PhantomJS in the first place is to generate PNGs. Since different browsers might have slightly different layout algorithms, using PhantomJS to create the PNG instead of the user's browser ensured that the pixels are reproducible.

However, all the PNG-generating code [seems to have been removed](https://github.com/Khan/graphie-to-png/commit/a248ec10827321b61984c59c221151345ee2e882) from this repo in favor of using SVGs. The Graphie data that gets saved to S3 when you click "convert to image" is just some SVG markup and the text/TeX of all the labels. No actual pixels are stored.

The sequence diagram for rendering to SVG with PhantomJS looked something like this:

with-phantomjs

The new one looks like this:

without-phantomjs

Source code for those sequence diagrams (generated with https://sequencediagram.org/)

```
title Rendering to SVG with PhantomJS

User's Browser->/svg endpoint: JS from the editor
/svg endpoint->PhantomJS:JS
PhantomJS->/svgize endpoint:JS
/svgize endpoint->PhantomJS:Graphie editor HTML with JS pre-populated
PhantomJS->PhantomJS:svg = window.getSVG()
PhantomJS->PhantomJS:other_data = window.getData()
PhantomJS->/svg endpoint:svg and other_data
/svg endpoint->S3:upload JSON containing svg and other_data
/svg endpoint->User's Browser:Graphie URL
```

```
title Rendering to SVG without PhantomJS

User's Browser->User's Browser:svg = window.getSVG()
User's Browser->User's Browser:other_data = window.getData()
User's Browser->/svg endpoint:JS, SVG, and other_data
/svg endpoint->S3:upload JSON containing svg and other_data
/svg endpoint->User's Browser:Graphie URL
```

**Comment by Myranae** [comment]
Wow, thanks so much for doing so much research into this! Love that sequence diagram website; so handy. I went to test it again though, and the little spinner just spins forever. I'm not sure I understand all the code involved with PhantomJS, but maybe the remaining code needs something from `_js_to_svg(js)` still? It looked like some code related to timing out also was deleted, so maybe that's why I'm not getting an error message this time? Totally just some slightly educated guesses, but up for pairing if you would like another perspective.

https://github.com/Khan/graphie-to-png/assets/60857422/e95c05c9-a3f5-4218-ac68-c82fe9a138f7

**Comment by benchristel** [comment]
Okay, I found (and fixed) another bug. I don't think it was related to the problem you were having @Myranae, but maybe it's worth re-testing? I've updated the test plan to cover the bug.

I think we probably shouldn't deploy this until January. It's too risky, with me being out, and not urgent.

**Review by Myranae — APPROVED** [review]
Thanks for pairing with me on this, Ben! Turns out some of the issues I experienced revolved around using the Dark Reader Chrome browser extension. We turned that off and found the main issue to be the secrets I was using were not the correct secrets. Ben had previously been able to obtain the secrets, so we paired through the testing phase for this code. It worked perfectly! Thanks for your work here! :)
